### PR TITLE
PR: Enclose python exe path in double quotes

### LINKED ISSUE #6746: 'Execute in an external system terminal' doesn't work if there is spaces in the path where the python interpreter is installed in Windows
## Problem Description

I'm trying to execute a script in an external system terminal on Windows 10 64bit and it is not working.

### What steps will reproduce the problem?
Let's say I have this very simple script :
```python
# -*- coding: utf-8 -*-
print('test')
```
1.  Run the script with <kbd>F5</kbd> using the configuration per file shown below so that the script is executed in an external system terminal and so that I can interact with the Python console after execution.

image

### What is the expected output? What do you see instead?

I expect the script to run in an external system terminal and to be able to see the result in the Python console. Instead, the cmd windows is shown for a fraction of a second, then nothing happens.

### Paste Traceback/Error Below (if applicable)

There is no traceback/error.

## Package Versions

* Spyder: Spyder 4.0.0.dev0 
* Python: Python 3.6.4 64bits
* Qt: Qt 5.9.1
* PyQt: PyQt5 5.9
* Operating System: Windows 10 amd64

### Dependencies

```
pyflakes >=0.6.0 :  1.6.0 (OK)
pycodestyle >=2.3:  2.3.1 (OK)
pygments >=2.0   :  2.2.0 (OK)
pandas >=0.13.1  :  0.22.0 (OK)
numpy >=1.7      :  1.14.2 (OK)
sphinx >=0.6.6   :  1.7.1 (OK)
rope >=0.9.4     :  0.10.7 (OK)
jedi >=0.9.0     :  0.11.1 (OK)
nbconvert >=4.0  :  5.3.1 (OK)
sympy >=0.7.3    :  1.1.1 (OK)
cython >=0.21    :  0.27.3 (OK)
qtconsole >=4.2.0:  4.4.0.dev (OK)
IPython >=4.0    :  6.2.1 (OK)
pylint >=0.25    :  1.8.2 (OK)
```

### PR DESCRIPTION
**Pull request by jnsebgosselin**
Fixes #6746

Enclose the path of the python executable in double quotes so that executing a script in an external system terminal works correctly when the path to the Python interpreter contains spaces.

### CODE REVIEW

File: spyder/utils/programs.py [hunk 1]
```diff
@@ -259,16 +259,19 @@ def run_python_script_in_terminal(fname, wdir, args, interact,
...
+
+    # If fname or python_exe contains spaces, it can't be ran on Windows, so we
+    # have to enclose them in quotes. Also wdir can come with / as os.sep, so
+    # we need to take care of it.
     if os.name == 'nt':
         fname = '"' + fname + '"'
         wdir = wdir.replace('/', '\\')
-    
-    p_args = [get_python_executable()]
+        python_exe = '"' + python_exe + '"'
```
Comment (csabella) [inline spyder/utils/programs.py]: Maybe enclose in triple quotes to not run into #6451?  I'm not sure if it's necessary.

Comment (ccordoba12) [reply spyder/utils/programs.py]: Good idea @csabella.

**Comment by ccordoba12** [comment]
@jnsebgosselin, the test failures are unrelated to this. I'll fix them in #6736.

File: spyder/utils/programs.py (hunk 1 above)
Comment (jnsebgosselin) [reply spyder/utils/programs.py]: Good point. I'll take that into consideration and push a commit with the changes.

**Comment by jnsebgosselin** [comment]
Ok, after this PR, I'm aiming to be able to execute this script
```
# -*- coding: utf-8 -*-
print('test')
```
in an external system console with a Python executable located in `C:\Python36's Path Test` in a file saved in `C:\Queen's Space\let's test this.py`.

Anything else I should include?

**Comment by ccordoba12** [comment]
Using utf-8 chars in the path is always a good idea (e.g. Chinese or Arabic chars).

**Comment by jnsebgosselin** [comment]
@ccordoba12 Have we defined a typical string to use to test this? If not, maybe we should so that our test for this are consistent across?

**Comment by ccordoba12** [comment]
There are some Chinese and Arabic chars floating around in several test files. Perhaps you could try to define them as a constant in a single place in this PR.

**Comment by jnsebgosselin** [comment]
> Perhaps you could try to define them as a constant in a single place in this PR.

Ok good idea, I'll try that.

**Comment by jnsebgosselin** [comment]
@ccordoba12 I'm setting the path to the interpreter in the `Python  interpreter` option as in the screenshot below.

image

It seems the IPython console is picking the right Python interpreter, since it throws an error about ipykernel and cloudpickle not being installed, which is expected since the Python env is empty.

image

But, when I'm running the script in a external console, it doesn't use the interpreter I've set in the option, it is using the Python interpreter used to run Spyder (I think).

Is it by design it is done that way? That would be more useful if the script was executed with the Python interpreter I've set in the option instead.

**Comment by jnsebgosselin** [comment]
This is the function used to fetch the path to the python executable located in `spyder\utils\misc.py` :

```python
def get_python_executable():
    """Return path to Python executable"""
    executable = sys.executable.replace("pythonw.exe", "python.exe")
    if executable.endswith("spyder.exe"):
        # py2exe distribution
        executable = "python.exe"
return executable
```

So I guess it indeed return the path to the Python executable in which Spyder is running?

**Comment by ccordoba12** [comment]
> So I guess it indeed return the path to the Python executable in which Spyder is running?

Yes. `sys.executable` returns the interpreter on which Spyder is running.

**Comment by jnsebgosselin** [comment]
> Yes. sys.executable returns the interpreter on which Spyder is running.

But why not using the Python interpreter that is set in the preferences?

**Comment by ccordoba12** [comment]
> But why not using the Python interpreter that is set in the preferences?

I have no problem with that. I think that's what most people would expect, but no one has complained about it before.

**Comment by jnsebgosselin** [comment]
> I have no problem with that. I think that's what most people would expect, but no one has complained about it before.

Can I implement this in this PR? If so, should this be done against master?

**Comment by ccordoba12** [comment]
> Can I implement this in this PR?

Yes, please do.

> If so, should this be done against master?

3.x is fine. I think this change is no disruptive enough to justify making it against master.

File: spyder/utils/programs.py [hunk 2]
```diff
@@ -253,22 +253,28 @@ def get_python_args(fname, python_args, interact, debug, end_args):
...
         fname = '"' + fname + '"'
         wdir = wdir.replace('/', '\\')
-    
-    p_args = [get_python_executable()]
+        executable = '"' + executable + '"'
+
+    p_args = [executable]
     p_args += get_python_args(fname, python_args, interact, debug, args)
     
+    print(executable)
```
Comment (CAM-Gerlach) [inline spyder/utils/programs.py]: This is the line causing problems in the test that no print functions were used in the code. If you're printing this for debugging, ``debug_print`` is what it recommends as an alternative.

Comment (jnsebgosselin) [reply spyder/utils/programs.py]: hehe yes I know, but thanks. I always forget to remove them...

Comment (CAM-Gerlach) [reply spyder/utils/programs.py]: I figured that was the other alternative, haha—happens to me too.

**Comment by ccordoba12** [comment]
@jnsebgosselin, adding a test for this is not easy, so I think this is ready.

**Comment by ccordoba12** [comment]
The failure in CircleCI is unrelated to this.

**Comment by jnsebgosselin** [comment]
Thanks a lot @ccordoba12 for reviewing and rounding up this PR!

**Comment by ccordoba12** [comment]
No problem :+1:

**Pull request merged**